file-attribute-caching/src/main/kotlin/com/pkware/file/fileAttributeCaching/FileAttributeCachingPath.kt

```diff
@@ -0,0 +1,211 @@
+package com.pkware.file.fileAttributeCaching
+
+import com.pkware.file.forwarding.ForwardingPath
+import java.nio.file.FileSystem
+import java.nio.file.Path
+import java.nio.file.attribute.BasicFileAttributes
+import java.nio.file.attribute.DosFileAttributes
+import java.nio.file.attribute.PosixFileAttributes
+import java.util.concurrent.TimeUnit
+import java.util.function.Function
+
+// Set the cache preservation duration to 5 seconds as most file operations complete in this time.
+private const val CACHE_PRESERVATION_SECONDS: Long = 5
+
+/**
+ * A [Path] instance that supports caching of [BasicFileAttributes] and other classes that extend it such as
+ * [DosFileAttributes] and [PosixFileAttributes].
+ *
+ * The cache duration is preset to be [CACHE_PRESERVATION_SECONDS].
+ *
+ * @param fileSystem the [FileSystem] associated with this [FileAttributeCachingPath] instance.
+ * @param delegate the [Path] to forward calls to if needed.
+ */
+internal class FileAttributeCachingPath(
+    private val fileSystem: FileSystem,
+    internal val delegate: Path
+) : ForwardingPath(delegate) {
+
+    // ExpirableCache for thisFileAttributeCachingPath a BasicFileAttributes cache
+    private val attributeCache = ExpirableCache<String, BasicFileAttributes?>(
+        TimeUnit.SECONDS.toMillis(CACHE_PRESERVATION_SECONDS)
+    )
+
+    override fun getFileSystem(): FileSystem = fileSystem
+
+    override fun resolve(other: Path): Path {
+        // Make sure if other is a FileAttributeCachingPath that we pass along other's
+        // delegate rather than other itself.
+        if (other is FileAttributeCachingPath) return delegate.resolve(other.delegate)
+        return delegate.resolve(other)
+    }
+
+    /**
+     * Sets the cache entry for the given attribute [name] with the given [value]. Can only set entire
+     * attribute `Class`es such as "dos:*", "posix:*", and "basic:*"
+     *
+     * The attribute name must include a "*" in order to be set within the cache.
+     *
+     * @param name The name of the attribute to cache.
+     * @param value The attribute value to cache.
+     */
+    fun <A : BasicFileAttributes?> setAttributeByName(name: String, value: A?) {
```

Comment:
this is because we are setting any attributes here that extend BasicFileAttributes, not just BasicFileAttributes only.

Reply:
so it would populate the cache with BasicFileAttributes, PosixFileAttributes, DosFileAttributes objects that are those specific types? because thats what I want, I do not want all BasicFileAttributes entires in the cache

Reply:
really hating that you cannot set a key or value in a map to a type that's extended by other types and it has to be explicit even when using generics

my ideal signature for the cache would be <String, <A : BasicFileAttributes?>> but thats not allowed.